plutus-core/plutus-ir/src/PlutusIR/Transform/Rename.hs

```diff
@@ -58,32 +64,235 @@ Two problems arise:
+-- where all mutually recursive data types (if any) are bound.
+renameConstrCM
+    :: (PLC.HasUniques (Term tyname name uni fun ann), PLC.MonadQuote m)
+    => Restorer PLC.ScopedRenaming m
+    -> VarDecl tyname name uni fun ann
+    -> ContT c (PLC.ScopedRenameT m) (PLC.ScopedRenameT m (VarDecl tyname name uni fun ann))
+renameConstrCM restorerAfterData (VarDecl ann name ty) = do
+    nameFr <- ContT $ PLC.withFreshenedName name
+    pure $ VarDecl ann nameFr <$> renameConstrTypeM restorerAfterData ty
+
+-- | Apply a function to a value in the non-recursive case and return the value unchanged otherwise.
+onNonRec :: Recursivity -> (a -> a) -> a -> a
+onNonRec NonRec f x = f x
+onNonRec Rec    _ x = x
+
+-- See Note [Weird IR data types]
+-- See Note [Renaming of constructors] (and make sure you understand all the intricacies in it
+-- before toucing this function).
 -- | Rename a 'Datatype' in the CPS-transformed 'ScopedRenameM' monad.
 renameDatatypeCM
     :: (PLC.HasUniques (Term tyname name uni fun ann), PLC.MonadQuote m)
-    => Datatype tyname name uni fun ann
+    => Recursivity
+    -> Datatype tyname name uni fun ann
     -> ContT c (PLC.ScopedRenameT m) (PLC.ScopedRenameT m (Datatype tyname name uni fun ann))
-renameDatatypeCM (Datatype x dataDecl params matchName constrs) = do
+renameDatatypeCM recy (Datatype x dataDecl params matchName constrs) = do
     -- The first stage (the data type itself, its constructors and its matcher get renamed).
-    dataDeclFr  <- ContT $ PLC.withFreshenedTyVarDecl dataDecl
-    constrsRen  <- traverse (ContT . PLC.withFreshenedVarDecl) constrs
-    matchNameFr <- ContT $ PLC.withFreshenedName matchName
+    -- Note that all of these are visible downstream.
+    restorerBeforeData <- captureContext
+    dataDeclFr         <- ContT $ PLC.withFreshenedTyVarDecl dataDecl
+    restorerAfterData  <- captureContext
+    constrsRen         <- traverse (renameConstrCM restorerAfterData) constrs
+    matchNameFr        <- ContT $ PLC.withFreshenedName matchName
     -- The second stage (the type parameters and types of constructors get renamed).
-    pure $
+    -- Note that parameters are only visible in the types of constructors and not downstream.
+    pure . onNonRec recy (unRestorer restorerBeforeData) $
         runContT (traverse (ContT . PLC.withFreshenedTyVarDecl) params) $ \paramsFr ->
             Datatype x dataDeclFr paramsFr matchNameFr <$> sequence constrsRen
 
--- | Rename a 'Binding' in the CPS-transformed 'ScopedRenameM' monad.
-renameBindingCM
+-- | Rename a 'Binding' from a non-recursive family in the CPS-transformed 'ScopedRenameM' monad.
+renameBindingNonRecC
+    :: (PLC.HasUniques (Term tyname name uni fun ann), PLC.MonadQuote m)
+    => Binding tyname name uni fun ann
+    -> ContT c (PLC.ScopedRenameT m) (Binding tyname name uni fun ann)
+renameBindingNonRecC binding = ContT $ \cont -> case binding of
```

Comment:
I originally wrote it this way. Here's how it looks like:

```haskell
renameBindingNonRecC
    :: (PLC.HasUniques (Term tyname name uni fun ann), PLC.MonadQuote m)
    => Binding tyname name uni fun ann
    -> ContT c (PLC.ScopedRenameT m) (Binding tyname name uni fun ann)
renameBindingNonRecC binding = case binding of
    TermBind x s var term -> do
        termFr <- lift $ renameTermM term
        varFr <- lift =<< ContT (PLC.withFreshenedVarDecl var)
        pure $ TermBind x s varFr termFr
    TypeBind x var ty -> do
        tyFr <- lift $ PLC.renameTypeM ty
        varFr <- ContT (PLC.withFreshenedTyVarDecl var)
        pure $ TypeBind x varFr tyFr
    DatatypeBind x datatype -> do
        datatypeFr <- lift =<< renameDatatypeCM NonRec datatype
        pure $ DatatypeBind x datatypeFr
```

but then you need to learn that `lift` means "perform the second stage right away instead of delaying it", so this is more concepts to learn. Using continuations directly is something that we do casually even in the PLC renamer with stuff like

```haskell
withFreshenedTyVarDecl
    :: (HasRenaming ren TypeUnique, HasUniques (Type tyname uni ann), MonadQuote m, MonadReader ren m)
    => TyVarDecl tyname ann
    -> (TyVarDecl tyname ann -> m c)
    -> m c
withFreshenedTyVarDecl (TyVarDecl ann name kind) cont =
    withFreshenedName name $ \nameFr -> cont $ TyVarDecl ann nameFr kind
```

so there's nothing new here. Anyway, I don't really care how it's written, so pick what you like more and I'll make it that.

Reply:
I mean, the current version is *also* very subtle in terms of when things get run. I didn't even notice the difference between the `TermBind` and the `TypeBind` cases. I'd at least put some comments highlighting those lines. The second version is perhaps even less comprehensible but at least those cases stand out :upside_down_face: 

I guess leave it but maybe add some more comments. Continuations are hard.